File: Leetcode/0001-1000/0001-0100/0002. Add Two Numbers.py
# https://leetcode.com/problems/add-two-numbers/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        preResult = ListNode()
        carry = 0
        iter1 = l1
        iter2 = l2
        iterResult = preResult
        while iter1 != None and iter2 != None:
            tempNode = ListNode((iter1.val + iter2.val + carry) % 10)
            carry = (iter1.val + iter2.val + carry) // 10
            iterResult.next = tempNode
            iterResult = iterResult.next
            iter1 = iter1.next
            iter2 = iter2.next
        
        while iter1 != None:
            tempNode = ListNode((iter1.val + carry) % 10)
            carry = (iter1.val + carry) // 10
            iterResult.next = tempNode
            iterResult = iterResult.next
            iter1 = iter1.next

        while iter2 != None:
            tempNode = ListNode((iter2.val + carry) % 10)
            carry = (iter2.val + carry) // 10
            iterResult.next = tempNode
            iterResult = iterResult.next
            iter2 = iter2.next

        if carry != 0:
            iterResult.next = ListNode(carry)
        
        return preResult.next

File: Leetcode/0001-1000/0001-0100/0002. Add Two Numbers.cpp
// https://leetcode.com/problems/add-two-numbers/

#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

static auto fast_io = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
}();

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
   public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode();
        int carry = 0;
        ListNode *it1 = l1, *it2 = l2, *it = head;
        while (it1 != nullptr && it2 != nullptr) {
            ListNode* temp = new ListNode((it1->val + it2->val + carry) % 10);
            carry = (it1->val + it2->val + carry) / 10;
            it->next = temp;
            it = it->next;
            it1 = it1->next;
            it2 = it2->next;
        }
        while (it1 != nullptr) {
            ListNode* temp = new ListNode((it1->val + carry) % 10);
            carry = (it1->val + carry) / 10;
            it->next = temp;
            it = it->next;
            it1 = it1->next;
        }
        while (it2 != nullptr) {
            ListNode* temp = new ListNode((it2->val + carry) % 10);
            carry = (it2->val + carry) / 10;
            it->next = temp;
            it = it->next;
            it2 = it2->next;
        }
        if (carry > 0) {
            it->next = new ListNode(carry);
        }
        return head->next;
    }
};